# RNA-Seq Workflow
Shaleigh Smith  

## Transcriptomic Data Processing 

### Downloading Data
- [SRA](https://www.ncbi.nlm.nih.gov/sra)

### Quality Control
- [FASTQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/) 

### Trimming 
- [TrimGalore](https://www.bioinformatics.babraham.ac.uk/projects/trim_galore/) 
- [Trimmomatic](http://www.usadellab.org/cms/index.php?page=trimmomatic) 

### Alignment 
- [STAR](https://github.com/alexdobin/STAR)
- [BBMap](https://jgi.doe.gov/data-and-tools/bbtools/bb-tools-user-guide/bbmap-guide/)
- [Bowtie2](https://github.com/BenLangmead/bowtie2)
- [BWA](https://github.com/lh3/bwa)
- [Tophat](http://ccb.jhu.edu/software/tophat/index.shtml)
- [Hisat2](http://daehwankimlab.github.io/hisat2/)
- [Salmon](https://salmon.readthedocs.io/en/latest/index.html)
- [Kallisto](https://github.com/pachterlab/kallisto)

### Quantification
- [Subread featureCounts](http://subread.sourceforge.net)
- [Tximport](https://github.com/mikelove/tximport)

### Helpful References
[Comparison of RNA-Seq workflows & analyses](https://www.nature.com/articles/s41598-020-76881-x)
> Corchete, L.A., Rojas, E.A., Alonso-LÃ³pez, D. et al. Systematic comparison and assessment of RNA-seq procedures for gene expression quantitative analysis. Sci Rep 10, 19737 (2020). https://doi.org/10.1038/s41598-020-76881-x

## Differential Expression Analysis 
- [DESeq2](https://bioconductor.org/packages/release/bioc/html/DESeq2.html)

### View html preview [here](https://htmlpreview.github.io/?https://raw.githubusercontent.com/sas1531/RNA-Seq-Workflow/main/rna_seq_workflow.html)
