---
title: "rna_seq_workflow"
author: "Shaleigh Smith"
date: "9/14/2021"
output: html_document
---

# RNA-Seq Workflow

---

## Transcriptomic Data processing
1. Data download from SRA.
2. Trimming and quality control with TrimGalore.
3. Alignment with Kallisto (example alignment for STAR and BBmap included).
4. Quantification of kallisto with tximport. 
5. Differential expression analysis with DESeq2.

---

## Data
### [Whole genome sequencing & RNA sequencing of human well differentiated liposarcoma](https://www.ncbi.nlm.nih.gov/bioproject/PRJNA751390)  
#### Select all samples for [SRA selector](https://www.ncbi.nlm.nih.gov/sra?linkname=bioproject_sra_all&from_uid=751390):  
- SRR15320001  
- SRR15320004  
- SRR15320005  
- SRR15320006  
- SRR15320007  
- SRR15320009  
- SRR15320010  
- SRR15320008  
- SRR15320012  
- SRR15320011  
- SRR15320002  

---

```{bash UCSC HG38, eval=FALSE}

# Download human genome
# https://support.illumina.com/sequencing/sequencing_software/igenome.html
cd ~/Desktop/
pwd
wget http://igenomes.illumina.com.s3-website-us-east-1.amazonaws.com/Homo_sapiens/UCSC/hg38/Homo_sapiens_UCSC_hg38.tar.gz
tar -zxvf Homo_sapiens_UCSC_hg38.tar.gz

```

```{bash STAR, eval=FALSE}

# GCC path if needed for install: /usr/local/Cellar/gcc/11.2.0/bin/g++-11

# If using STAR for sequence alignment...
# Can only do this and map with 30GB+ of RAM - must use cluster

## Create STAR genome index
STAR --runThreadN 1 \
--runMode genomeGenerate \
--genomeDir ~/Desktop/work/rna_seq/data/star_hg38_index \
--genomeFastaFiles ~/Desktop/work/rna_seq/data/Homo_sapiens/UCSC/hg38/Sequence/WholeGenomeFasta/genome.fa \
--sjdbGTFfile ~/Desktop/work/rna_seq/data/Homo_sapiens/UCSC/hg38/Annotation/Genes/genes.gtf \
--sjdbOverhang 99

# Iterate through project Run IDs
cd ./data/fastq/
pwd
while IFS= read -r i
  do
  echo $i
  
  # Get SRA IDs for Liposarcoma samples
  # Download data from SRA, put each read into separate files (paired data)
  /Users/sha/sratoolkit.2.11.1-mac64/bin/fastq-dump --accession $i --split-files --gzip -O ./
  
  # Can also download manually here:
  # https://www.ebi.ac.uk/ena/browser/view/SRR15320006?show=reads
  
  # Trim and run FASTQC
  echo ${i}_1.fastq.gz
  echo ${i}_2.fastq.gz
  ~/TrimGalore-0.6.6/trim_galore --path_to_cutadapt  ~/.local/bin/cutadapt --q 30  --phred33  -o ./  --paired  --fastqc  ./${i}_1.fastq.gz  ./${i}_2.fastq.gz
  
  # Align reads to reference genome with STAR (on cluster)
  # Spliced Transcripts Alignment to a Reference (STAR) is a fast RNA-seq read mapper...
  # with support for splice-junction and fusion read detection.
  # https://github.com/alexdobin/STAR/blob/master/doc/STARmanual.pdf
  # MAPQ=255 is defaut (uniquely mapping reads)
  # Convert sam to bam and sort
  STAR --genomeDir ~/Desktop/work/rna_seq/data/star_hg38_index/ \
  --runThreadN 1 \
  --readFilesIn ${i}_1_val_1.fq.gz ${i}_2_val_2.fq.gz \
  --outSAMtype BAM SortedByCoordinate \
  --outSAMunmapped Within
  
  # Create a feature counts matrix
  # The input is specified as reversely stranded (-s 2) and paired end (-p). 
  # The fragment length is checked (-P) 
  # Only the fragments that have both ends successfully aligned are counted (-B)
  # Chimeric fragments are ignored (-C) as are duplicates (--ignoreDup)
  # Only primary alignments are counted (--primary) 
  # Counts are based on gene_id (-g).
  /Users/sha/subread-2.0.3-source/bin/featureCounts -s 2 -p -B -C -P  --ignoreDup  --primary  -a  ../Homo_sapiens_UCSC_hg38/Homo_sapiens/UCSC/hg38/Annotation/Genes.gencode/genes.gtf  -g gene_id  -o ../count/${i} -feature-counts ../align/${i}_sorted.bam
  
done < ../SRR_Acc_List.txt 

```

```{bash BBMap, eval=FALSE}

# Build BBMap index
/Users/sha/bbmap/bbmap.sh -Xmx20G ref=../Homo_sapiens_UCSC_hg38/Homo_sapiens/UCSC/hg38/Sequence/WholeGenomeFasta/genome.fa

# Iterate through project Run IDs
cd ./data/fastq/
pwd
while IFS= read -r i
  do
  echo $i
  
  # Get SRA IDs for Liposarcoma samples
  # Download data from SRA, put each read into separate files (paired data)
  /Users/sha/sratoolkit.2.11.1-mac64/bin/fastq-dump --accession $i --split-files --gzip -O ./
  
  # Can also download manually here:
  # https://www.ebi.ac.uk/ena/browser/view/SRR15320006?show=reads
  
  # Trim and run FASTQC
  echo ${i}_1.fastq.gz
  echo ${i}_2.fastq.gz
  ~/TrimGalore-0.6.6/trim_galore --path_to_cutadapt  ~/.local/bin/cutadapt --q 30  --phred33  -o ./  --paired  --fastqc  ./${i}_1.fastq.gz  ./${i}_2.fastq.gz
  
  # Align read to reference genome with BBMap
  # BBmap is splice-aware but uses less RAM than STAR
  # minid=0.76: Approximate minimum alignment identity to look for. Higher is faster and less sensitive.
  # ambiguous: Set behavior on ambiguously-mapped reads (with multiple top-scoring mapping locations).
  bbmap.sh -Xmx6G path=../bbmap_index/ in=SRR15320001_1_val_1.fq.gz  in2=SRR15320001_2_val_2.fq.gz  outm=../align/SRR15320001.sam  ambiguous=random 
  
  # Convert SAM file to BAM file and sort
  samtools view -S -b ../align/${i}.sam > ../align/${i}.bam
  samtools sort ../align/${i}.bam -o ../align/${i}_sorted.bam
  
  # Create a feature counts matrix
  # The input is specified as reversely stranded (-s 2) and paired end (-p). 
  # The fragment length is checked (-P) 
  # Only the fragments that have both ends successfully aligned are counted (-B)
  # Chimeric fragments are ignored (-C) as are duplicates (--ignoreDup)
  # Only primary alignments are counted (--primary) 
  # Counts are based on gene_id (-g).
  /Users/sha/subread-2.0.3-source/bin/featureCounts -s 2 -p -B -C -P  --ignoreDup  --primary  -a  ../Homo_sapiens_UCSC_hg38/Homo_sapiens/UCSC/hg38/Annotation/Genes.gencode/genes.gtf  -g gene_id  -o ../count/${i} -feature-counts ../align/${i}_sorted.bam
  
done < ../SRR_Acc_List.txt 

```

```{bash Kallisto - Data download, trimming, QC, and pseudoalignment on local, eval=FALSE}

# Create kallisto index
# Must build from homo sapiens coding sequences (transcriptome of organism of interest)
# Can download from Ensembl FTP Site: http://useast.ensembl.org/info/data/ftp/index.html 
kallisto index --index=Homo_sapiens_GRCh38_cds_all_kallisto_index Homo_sapiens.GRCh38.cds.all.fa

# Iterate through project Run IDs
cd ./data/fastq/
pwd
while IFS= read -r i
  do
  echo $i
  
  # Get SRA IDs for Liposarcoma samples
  # Download data from SRA, put each read into separate files (paired data)
  /Users/sha/sratoolkit.2.11.1-mac64/bin/fastq-dump --accession $i --split-files --gzip -O ./
  
  # Can also download manually here:
  # https://www.ebi.ac.uk/ena/browser/view/SRR15320006?show=reads
  
  # Trim and run FASTQC
  echo ${i}_1.fastq.gz
  echo ${i}_2.fastq.gz
  ~/TrimGalore-0.6.6/trim_galore --path_to_cutadapt  ~/.local/bin/cutadapt --q 30  --phred33  -o ./  --paired  --fastqc  ./${i}_1.fastq.gz  ./${i}_2.fastq.gz
  
  # Run kallisto
  # Get abundance estimates through pseudoalignment 
  kallisto quant -i ../kallisto/Homo_sapiens_GRCh38_cds_all_kallisto_index -o ../align/${i} -b 100 --bias ${i}_1_val_1.fq.gz ${i}_2_val_2.fq.gz
  
done < ../SRR_Acc_List.txt 

```


---

# Differential Expression Analysis

```{r Library, message=FALSE}

# Library
library(tidyverse)
library(biomaRt)
library(tximport)
library(rhdf5)
library(DESeq2)
library(ComplexHeatmap)
library(circlize)
library(AnnotationDbi)
library(org.Hs.eg.db)
library(ggrepel)

```

```{r Kallisto preprocessing}

# Import ensembla annotations for the human genome
mart <- biomaRt::useMart(biomart = "ensembl", dataset =  "hsapiens_gene_ensembl")
bio_mart_df <- biomaRt::getBM(mart = mart,
                         attributes = c("ensembl_transcript_id", "transcript_version", "ensembl_gene_id",
                                        "external_gene_name", "description", "transcript_biotype",
                                        "gene_biotype"))

# Combine transcript id and version
bio_df <- bio_mart_df %>%
  mutate(target_id = paste0(ensembl_transcript_id, ".", transcript_version)) %>%
  dplyr::rename(gene_symbol = external_gene_name, 
                full_name = description, 
                biotype = transcript_biotype) %>%
  dplyr::select(-ensembl_transcript_id, -transcript_version)
head(bio_df)

# Convert kallisto abundance.tsv (or abundance.h5) files into a gene count format for analysis using either DESeq2
# Use tximport to summarize transcript counts into gene counts
kallisto_names <- list.dirs("/Users/sha/Desktop/work/rna_seq/data/align/",  full.names = FALSE)[-1]
kallisto_dir <- list.dirs("/Users/sha/Desktop/work/rna_seq/data/align/", full.names = TRUE)[-1]
kallisto_files <- file.path(kallisto_dir,"abundance.tsv")
names(kallisto_files) <- kallisto_names
tx_df <- tximport(kallisto_files, 
                  type = "kallisto",
                  tx2gene = dplyr::select(bio_df, target_id, gene_symbol),
                  countsFromAbundance ="no")
summary(tx_df)

# Subset output for protein coding genes only 
# Round all values (DESEQ does not like fractions/decimals)
count_df <- tx_df$counts %>%
  as.data.frame() %>%
  tibble::rownames_to_column("gene_symbol") %>%
  filter(gene_symbol %in% filter(bio_df, gene_biotype == "protein_coding")$gene_symbol &
           gene_symbol != "") %>% 
  mutate_if(is.numeric, round) %>%
  tibble::column_to_rownames("gene_symbol")
head(count_df)

# Save
#write.table(count_df, "./data/count/kallisto_counts.txt", sep = "\t", row.names = T, quote = F)

```

```{r Kallisto - Create DESeq2 data set}

# Create sample annotation df for conditions input
clin_df <- read.delim("./data/SraRunTable.txt", sep = ",")
clin_df <- clin_df %>%
  dplyr::select(Run, tissue) %>%
  arrange(Run) %>% 
  mutate_if(is.character, ~ gsub(" ", "_", .)) %>%
  mutate(tissue = factor(tissue, levels = c("Normal_fat","Liposarcoma")))
clin_df

# Read in count data
count_df <- read.delim("./data/count/kallisto_counts.txt")

# Check order
clin_df$Run == colnames(count_df) #TRUE

# Create DESeq2 data set
de_data_df <- DESeqDataSetFromMatrix(countData = count_df, 
                                     colData = clin_df, 
                                     design = ~ tissue)
head(de_data_df)

# Remove genes that do not have counts greater than 2 in at least 2 of the samples (columns)
dim(de_data_df)
de_data_df <- de_data_df[rowSums(counts(de_data_df) >= 2) >= 2] 
dim(de_data_df)

# Confirm that all samples are labelled correctly 
as.data.frame(colData(de_data_df))

```

```{r DESeq2 matrix EDA normalization, warning=FALSE}

### Exploratory data analysis of DESeq2 matrix with transformation 

# Estimate size factors
# The size factor is the median ratio of the sample over a pseudosample: for each gene, the geometric mean of all samples
deseq_eda <- estimateSizeFactors(de_data_df)

# Apply regularized-logarithm transformation
rld <- rlog(deseq_eda, blind = FALSE)

# Apply variance stabilizing transformation
vsd <- vst(deseq_eda, blind = FALSE)

# Create new data frame three normalization methods for all samples
deseq_eda <- bind_rows(
  as_data_frame(log2(counts(deseq_eda, normalized=TRUE)[, (1:8)])) %>%
         mutate(transform = "log2(x + 1)"),
  as_data_frame(assay(vsd)[, (1:8)]) %>% mutate(transform = "vst"),
  as_data_frame(assay(rld)[, (1:8)]) %>% mutate(transform = "rlog"))
  
# Compare transformation visually 
ggplot(deseq_eda, aes(x = SRR15320001, y = SRR15320002)) + 
  geom_bin2d(bins = 40) +
  coord_fixed() + 
  facet_grid( . ~ transform) +
  scale_fill_continuous(limits = c(0, 2000), breaks = seq(0, 2000, by = 500)) +
  ggtitle("SRR15320001 vs. SRR15320002")+
  theme_classic() + 
  theme(plot.title = element_text(hjust = 0.5),
        panel.grid.major = element_line()) 

ggplot(deseq_eda, aes(x = SRR15320001, y = SRR15320004)) + 
  geom_bin2d(bins = 40) +
  coord_fixed() + 
  facet_grid( . ~ transform) +
  scale_fill_continuous(limits = c(0, 2000), breaks = seq(0, 2000, by = 500)) +
  ggtitle("SRR15320001 vs. SRR15320004") +
  theme_classic() + 
  theme(plot.title = element_text(hjust = 0.5),
        panel.grid.major = element_line()) 

ggplot(deseq_eda, aes(x = SRR15320002, y = SRR15320006)) + 
  geom_bin2d(bins = 40) +
  coord_fixed() + 
  facet_grid( . ~ transform) +
  scale_fill_continuous(limits = c(0, 2000), breaks = seq(0, 2000, by = 500)) +
  ggtitle("SRR15320002 vs. SRR15320006") +
  theme_classic() + 
  theme(plot.title = element_text(hjust = 0.5),
        panel.grid.major = element_line()) 


```

```{r DESeq2 matrix EDA distance}

### Exploratory data analysis of DESeq matrix with sample comparison 
# This will asses overall similarity between samples
# As shown in the above plots this similarity might not be as expected

# Calculate the euclidean distance between samples
deseq_distance <- dist(t(assay(rld)))
deseq_distance

# Create annotation for the heatmap
deseq_ha <- HeatmapAnnotation(df = dplyr::select(clin_df, tissue),
                              col = list(tissue = c("Normal_fat" = "darkred",
                                                    "Liposarcoma" = "darkblue")))

# Visualize with heatmap 
deseq_heat <- Heatmap(as.matrix(deseq_distance),
                   top_annotation = deseq_ha,
                   show_column_names = F, 
                   show_row_names = T,
                   cluster_rows = T,
                   cluster_columns = T,
                   clustering_method_columns = 'complete',
                   clustering_method_rows = "complete",
                   row_names_gp = gpar(fontsize = 8),
                   heatmap_legend_param = list(title = "Distance"),
                   col = colorRamp2(c(min(deseq_distance), max(deseq_distance)), c("darkorchid4", "white")))

deseq_heat

```

```{r DESeq2 matrix EDA PCA}

### Compare with PCA 
# Create PCA object to be plotted with ggplot
plot_pca <- plotPCA(rld, intgroup = c("tissue"), returnData = TRUE)

# Calculate and round variance for plotting
percentVar <- round(100 * attr(plot_pca, "percentVar"))

# Plot PCA           
deseq_pca_1 <- ggplot(plot_pca, aes(x = PC1, y = PC2, color = tissue)) +
  geom_point(size = 2.5) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  coord_fixed() +
  scale_color_manual(values = c("darkred", "darkblue")) + 
  ggtitle("DESeq2 PCA") +
  theme_classic() +
  theme(plot.title = element_text(hjust = 0.5),
        panel.grid.major = element_line()) 

deseq_pca_1

```

```{r DESeq2 run analysis}

# Make sure that normal is the first level in the sample factor
# This is to ensure the log2 fold change is calculated over the control (when results are called at random)
levels(de_data_df$tissue)

# Check to insure all samples are correct 
as.data.frame(colData(de_data_df))

# Run the DESeq analysis against raw counts 
de_final_df <- DESeq(de_data_df)
results(de_final_df)

# Call results
# Contrast: a character vector with exactly three elements: 
# 1. the name of a factor in the design formula
# 2. the name of the numerator level for the fold change
# 3. the name of the denominator level for the fold change (simplest case)
de_res_df <- results(de_final_df, contrast = c("tissue", "Liposarcoma", "Normal_fat"))
de_res_df

# Review Comparisons 
mcols(de_res_df, use.names = TRUE)

### Review Summary
summary(de_res_df, alpha = 0.01)

# Save
write.table(as.data.frame(de_res_df) %>% tibble::rownames_to_column("gene_symbol"),
            "./deseq2_liposarcoma_results.txt", sep = "\t", row.names = F, quote = F)

```

```{r DESeq2 significance wald test, warning=FALSE}

# Visualize significance
# Use adjusted p value threshold of 0.1 due to low number of significanlty regulated genes
# Also indicate genes of interest
vol_df <- as.data.frame(de_res_df) %>%
  tibble::rownames_to_column("gene_symbol") %>%
  mutate(threshold = case_when(log2FoldChange > 0 & padj < 0.1 ~ "up_sig",
                               log2FoldChange < 0 & padj < 0.1 ~ "down_sig",
                               TRUE ~ "not_sig"))

# Volcano plot
vol_plot <- ggplot(vol_df, aes(x = log2FoldChange, y = -log10(pvalue))) +
  geom_point(aes(col = threshold), size = 1, alpha = 0.9) +
  ggtitle("Liposarcoma vs. Normal Fat") + 
  theme_classic() +
  theme(plot.title = element_text(hjust = 0.5),
        panel.grid.major = element_line()) +
  scale_colour_manual(values = c("red3", "grey", "blue", "black"), 
                       breaks = c("up_sig", "not_sig", "down_sig", "other"),
                       labels = c("Up", "N.S.", "Down", "Other")) +
  geom_text_repel(data = filter(vol_df, padj < 0.1 | gene_symbol %in% c("CDK4", "MDM2", "PPARG", "CEBPA")),
                  aes(label = gene_symbol), size = 2) +
  labs(col = "Significance\np.adj < 0.1")
vol_plot

```

```{r DESeq2 shrinkage of logFC estimates, warning=FALSE}

# Apply shrinkage to logFC 
# See difference in visualization

# Calculate row means across samples 
mean_norm <-  rowMeans(counts(de_final_df, normalized=TRUE)[, de_final_df$tissue == "Liposarcoma"])
mean_tumor <-  rowMeans(counts(de_final_df, normalized=TRUE)[, de_final_df$tissue == "Normal_fat"])

### Normal vs. Tumor
# Log Fold change
lfc_comp <- lfcShrink(de_final_df, coef = "tissue_Liposarcoma_vs_Normal_fat", type = "apeglm")

# Add row names to log fold change data frame 
lfc_comp <-  cbind(as.data.frame(lfc_comp), mean_norm, mean_tumor)

# Visualize significance
# Use adjusted p value threshold of 0.1 due to low number of significanlty regulated genes
# Also indicate genes of interest
lfc_vol_df <- lfc_comp %>%
  tibble::rownames_to_column("gene_symbol") %>%
  mutate(threshold = case_when(log2FoldChange > 0 & padj < 0.1 ~ "up_sig",
                               log2FoldChange < 0 & padj < 0.1 ~ "down_sig",
                               TRUE ~ "not_sig"))

# Volcano plot
lfc_vol_plot <- ggplot(lfc_vol_df, aes(x = log2FoldChange, y = -log10(pvalue))) +
  geom_point(aes(col = threshold), size = 1, alpha = 0.9) +
  ggtitle("Liposarcoma vs. Normal Fat") + 
  theme_classic() +
  theme(plot.title = element_text(hjust = 0.5),
        panel.grid.major = element_line()) +
  scale_colour_manual(values = c("red3", "grey", "blue", "black"), 
                       breaks = c("up_sig", "not_sig", "down_sig", "other"),
                       labels = c("Up", "N.S.", "Down", "Other")) +
  geom_text_repel(data = filter(lfc_vol_df, padj < 0.1 | 
                                  gene_symbol %in% c("CDK4", "MDM2", "PPARG", "CEBPA") |
                                  log2FoldChange > 2),
                  aes(label = gene_symbol), size = 2) +
  labs(col = "Significance\np.adj < 0.1")
lfc_vol_plot

```


---

